enhancements/api-review/add-new-CRD-ImageContentPolicy(ICP)-to-config.openshift.io.md

```diff
@@ -0,0 +1,325 @@
+
+The user need to define multiple ImageContentPolicies, used by apps/manifests that don't use digests to pull
+the images when working with disconnected environments or pulling from registries that act as transparent pull-through proxy cache.
+
+The user can pull image without digest by configuring mirrors using `mirrorByTags` in the ImageContentPolicy file.
+And MCO will the ImageContentPolicy project. Once this is done, the images can be pulled from the mirrors without the digest referenced.
+
+### API Extensions
+- Adds a CRD ImageContentPolicy to config.openshift.io/v1.
+- Impacts on existing resources(The impact of ImageContentPolicy in the "Operational Aspects
+of API Extensions" section)
+  - Components use ImageContentSourcePolicy should upgrade to use ImageContentPolicy instead of ImageContentSourcePolicy.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Implementing this enhancement requires changes in:
+- [containers/images](https://github.com/containers/image): registries.conf needs change its mirror setting to support separated mirror lists: a list of mirrors that allow for tag specification,
+and a list of mirrors require digest specification.
+- [openshift/api/config/v1](https://github.com/openshift/api/tree/master/config/v1): definition the schema and API of the ImageContentPolicy CRD.
+- [openshift/client-go](https://github.com/openshift/client-go), [openshift/cluster-config-operator](https://github.com/openshift/cluster-config-operator/pull/220): rebase the openshift/api version in these repositories to apply the new CRD
+to the cluster.
+- [openshift/runtime-utils/pkg/registries](https://github.com/openshift/runtime-utils/tree/master/pkg/registries): helper functions to edit registries.conf.
+- [openshift/machine-config-operator](https://github.com/openshift/machine-config-operator): MCO needs watch for the ImageContentPolicy.
+- This [document](https://docs.google.com/document/d/11FJPpIYAQLj5EcYiJtbi_bNkAcJa2hCLV63WvoDsrcQ/edit?usp=sharing) keeps a list of components that use operator.openshift.io/v1alpha1 ImageContentSourcePolicy. Need to change those repositories to upgrade to ImageContentPolicy.
+
+#### Details
+
+The merge order of mirrors for the same source:<br/>
+Order is preserving the relative order of the mirrors using topological sorting. A graph is formed using each mirror as a
+vertex, adding an unidirectional edge between adjacent mirrors from the same mirror list. If the graph is cyclic graph, the topological sort will start
+from the first vertex after the alphabetical sorting of the vertces. If multiple verteces is from the same vertex, they will be
+visited in alphabetical order.
+
+#### Notes
+
+1. During the upgrade path, MCO can watch for both ImageContentPolicy and ImageContentSourcePolicy and create objects.
+
+2. The ImageStream has not supported pulling using tags yet. It has started working on this feature.
+
+### Risks and Mitigations
```

Comment:
any implications for signature verification?

Reply:
what kind of implication do you mean? Openshift uses signature verification, it is still useful to make sure the images pulled by tags from the mirrors are trusted.

Reply:
when you mirror by SHA, you can verify the original content in the original registry, and you know because the mirrored SHA is the same, it's the same content.

when you mirror by tag, are you actually still able to verify the mirrored content is valid?  the signature won't match because the registry hostname is different, right?  And verifying the original source is useless because there's no guarantee the mirrored content matches the original source.

So i'm wondering what a user who cares about signature validation is expected to do, if they are using this feature?

Reply:
>  the signature won't match because the registry hostname is different, right?

No; mirrors are transparent to everything, incl. image names in pod specs, image list outputs, and the name check in signature verification.

The user just needs to mirror signatures along with the images (and possibly configure sigstore URLs for the mirrors on the nodes, if the mirror registry doesn’t have native X-R-S-S support).

Reply:
(Also, signatures are the primary mechanism that prevents malicious users from pointing tags to completely unexpected images; they contain a signed (repo:tag, digest) binding. (If the trusted image author signed two different images for the same tag, an attacker could perhaps switch between the two.))